cmd/clusterctl/pkg/client/config/providers_client.go

```diff
@@ -0,0 +1,196 @@
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+		&provider{
+			name:         "vsphere",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/releases/latest/infrastructure-components.yaml",
+			providerType: clusterctlv1.InfrastructureProviderType,
+		},
+
+		// Bootstrap providersClient
+		// TODO: CABPK in v1alpha3 will be included into CAPI, so this entry can be removed as soon as v1alpha3 is ready for test
+		&provider{
+			name:         "kubeadm",
+			url:          "https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/releases/latest/bootstrap-components.yaml",
+			providerType: clusterctlv1.BootstrapProviderType,
+		},
+	}
+
+	return defaults
+}
+
+// configProvider mirrors config.Provider interface and allows serialization of the corresponding info
+type configProvider struct {
+	Name string                    `json:"name,omitempty"`
+	URL  string                    `json:"url,omitempty"`
+	Type clusterctlv1.ProviderType `json:"type,omitempty"`
+}
+
+func (p *providersClient) List() ([]Provider, error) {
+	// Creates a maps with all the defaults provider configurations
+	providers := p.defaults()
+
+	// Gets user defined provider configurations, validate them, and merges with
+	// hard-coded configurations handling conflicts (user defined take precedence on hard-coded)
+
+	var userDefinedProviders []configProvider //nolint
+	if err := p.reader.UnmarshalKey(ProvidersConfigKey, &userDefinedProviders); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal providers from the clusterctl configuration file")
+	}
+
+	for _, u := range userDefinedProviders {
+		provider := NewProvider(u.Name, u.URL, u.Type)
+		if err := validateProviderRepository(provider); err != nil {
+			return nil, errors.Wrapf(err, "error validating configuration from %q. Please fix the providers value in clusterctl configuration file", provider.Name())
+		}
+
+		override := false
+		for i := range providers {
+			if providers[i].Name() == provider.Name() {
+				providers[i] = provider
+				override = true
+				klog.V(3).Infof("The clusterctl configuration file overrides default configuration for provider %s", provider.Name())
+			}
+		}
+
+		if !override {
+			providers = append(providers, provider)
+		}
+	}
+
+	// ensure provider configurations are consistently sorted
+	sort.Slice(providers, func(i, j int) bool {
+		return providers[i].Name() < providers[j].Name()
+	})
+
+	return providers, nil
+}
+
+func (p *providersClient) Get(name string) (*Provider, error) {
+	l, err := p.List()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, r := range l {
+		if name == r.Name() {
+			return &r, nil
+		}
+	}
+
+	return nil, errors.Errorf("failed to get configuration for %q provider. Please check the provider name and/or add configuration for new providers using the .clusterctl config file", name)
+}
+
+func validateProviderRepository(r Provider) error {
```

Comment:
nit: how about changing the name to just `validateProvider()` since you're doing several things - validating Type, name and URL.

Reply:
done